builder/digest_remote.go

```diff
@@ -0,0 +1,92 @@
+package builder
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/Azure/acr-builder/graph"
+	"github.com/Azure/acr-builder/pkg/image"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/docker/distribution/reference"
+	"github.com/pkg/errors"
+)
+
+type remoteDigest struct {
+	registryCreds graph.RegistryLoginCredentials
+}
+
+func NewRemoteDigest(creds graph.RegistryLoginCredentials) *remoteDigest {
+	return &remoteDigest{
+		registryCreds: creds,
+	}
+}
+
+var _ DigestHelper = &remoteDigest{}
+
+func (d *remoteDigest) PopulateDigest(ctx context.Context, ref *image.Reference) error {
+	if ref == nil {
+		return nil
+	}
+	if ref.Digest != "" {
+		return nil
+	}
+	client := http.DefaultClient
+	opts := docker.ResolverOptions{
+		Client: client,
+	}
+	if cred, ok := d.registryCreds[ref.Registry]; ok {
+		if cred.Username.ResolvedValue == "" || cred.Password.ResolvedValue == "" {
+			return fmt.Errorf("Error fetching credentials for '%s'", ref.Registry)
+		}
+		// Adds credential resolver if private registry
+		opts.Credentials = func(hostName string) (string, string, error) {
+			return cred.Username.ResolvedValue, cred.Password.ResolvedValue, nil
+		}
+	}
+
+	resolver := docker.NewResolver(opts)
+	imageRef, err := getReferencePath(ref)
+	if err != nil {
+		return err
+	}
+	// Note: (bug), if the final image is NOT pushed in this build,
+	// Then the `Resolve` command below will resolve the digest of the LAST pushed image with the same tag (IFF tag is provided)
+	// Then, the unwanted behavior of runtime dependencies found in THIS run being tracked on OLD image occurs. 
+	// For ex., consider following step:
+	// - id: build
+	//   build: -t $Registry/acb:latest -f Dockerfile .
+	//   env: ["DOCKER_BUILDKIT=1"]
+	// 
+	// Since we do not push the final image "acb:latest", `Resolve` will get digest of the "acb:latest" IFF it was ever pushed before. 
+	// Lets assume old digest was sha256:xyz
+	// Then, "$Registry/acb:latest" with digest "sha256:xyz" will start to track dependencies from THIS run.
+	// But, I assume MOST tasks always Push after Build, so the dependencies are tracked on the last Pushed digest, so it won't be an issue in MOST cases.
+	// TODO: We need to skip attaching the digest to the final image if the final image wasn't pushed in this Task run.
+	_, desc, err := resolver.Resolve(ctx, imageRef)
+	if err == nil {
+		ref.Digest = desc.Digest.String()
+		return nil
+	}
+	// If the final image is not pushed ever, it will not have any digest.
+	if errdefs.IsNotFound(err) {
+		return nil
+	}
+
+	return errors.Wrapf(err, "Failed to Resolve the reference '%s'", ref.Reference)
+}
+
+func getReferencePath(ref *image.Reference) (string, error) {
```

Comment:
In the docker inspect solution, we need to special handle library images, eg, nginx, ubuntu, because of the `library` prefix. Does the remote inspect work for these library image?

Reply:
That only needs to be done because the `docker inspect` returns an output like:
```
["busybox@sha256:b5fc1d7b2e4ea86a06b0cf88de915a2c43a99a00b6b3c0af731e5f4c07ae8eff","busybox@sha256:f3cfc9d0dbf931d3db4685ec659b7ac68e2a578219da4aae65427886e649b06b"]
```

We just remove the prefix library/ if we attached it.

We don't need to do any of that in remote